src/lib/triggersManager.js

```diff
@@ -0,0 +1,188 @@
+import SDK from '../api';
+import { store } from '../store';
+import { insert, createToken } from 'components/helpers';
+
+const agentCacheExpiry = 3600000;
+let agentPromise;
+const getAgent = (triggerAction) => {
+	if (agentPromise) {
+		return agentPromise;
+	}
+
+	agentPromise = new Promise(async (resolve, reject) => {
+		const { params } = triggerAction;
+
+		if (params.sender === 'queue') {
+			const { state } = store;
+			const { token, triggerAgent } = state;
+
+			if (triggerAgent) {
+				const cacheAgent = triggerAgent;
+
+				// cache valid for 1h
+				if (cacheAgent.ts && Date.now() - cacheAgent.ts < agentCacheExpiry) {
+					return resolve(cacheAgent.agent);
+				}
+			}
+
+			let agent;
+			try {
+				({ agent } = await SDK.nextAgent({ token }));
+			} catch (error) {
+				return reject(error);
+			}
+
+			store.setState({ 'triggerAgent': { agent, ts: Date.now() } });
+			resolve(agent);
+		} else if (params.sender === 'custom') {
+			resolve({
+				username: params.name,
+			});
+		} else {
+			reject('Unknown sender');
+		}
+	});
+
+	// expire the promise cache as well
+	setTimeout(() => {
+		agentPromise = null;
+	}, agentCacheExpiry);
+
+	return agentPromise;
+}
+
+class TriggersManager {
+	constructor() {
+		if (!TriggersManager.instance) {
+			this.started = false;
+			this.requests = [];
+			this.triggers = [];
+			this.enabled = true;
+			TriggersManager.instance = this;
+		}
+
+		return TriggersManager.instance;
+	}
+
+	start() {
+		if (this.started) {
+			return;
+		}
+
+		const { state } = store;
+		const { token, firedTriggers = [], config: { triggers } } = state;
+		SDK.credentials.token = token;
+
+		if (!(triggers && triggers.length > 0)) {
+			return;
+		}
+
+		this.started = true;
+		this.triggers = [...triggers];
+
+		firedTriggers.forEach((triggerId) => {
+			this.triggers.forEach((trigger) => {
+				if (trigger._id === triggerId) {
+					trigger.skip = true;
+				}
+			});
+		});
+
+		this.processTriggers();
+	}
+
+	fire(trigger) {
+		const { state } = store;
+		const { token, user, firedTriggers = [] } = state;
+		if (!this.enabled || user) { // need to think about testing user obj here..
+			return;
+		}
+
+		trigger.actions.forEach((action) => {
+			if (action.name === 'send-message') {
+				trigger.skip = true;
+
+				getAgent(action).then(async (agent) => {
+					const message = {
+						msg: action.params.msg,
+						token,
+						u: agent,
+						ts: new Date(),
+						_id: createToken(),
+					}
+
+					store.setState({ triggered: true, messages: insert(store.state.messages, message).filter(({ msg, attachments }) => ({ msg, attachments })) });
+
+					// TODO: Need to think about the implementation below.. Is it possible that when the room is created, the available agent is not the same one that was previously selected?
```

Comment:
Yes ... when the trigger action requests a valid agent, the method in the SDK will call the next available agent, but this action does not assign the agent to the room because the room has not yet been created. Therefore, if the guest takes a while to respond, when the room is created it is possible that the available agent is another.